cpp/src/arrow/dataset/file_base.cc

```diff
@@ -148,15 +160,20 @@ FragmentIterator FileSystemDataset::GetFragmentsImpl(
   FragmentVector fragments;
   std::vector<std::shared_ptr<ScanOptions>> options(forest_.size());
 
+  ExpressionVector fragment_partitions(forest_.size());
```

Comment:
I disagree; as previously discussed including the partition information of every parent is redundant and breaks the efficiency of pruning directories (since a parent directory's partition expression will be evaluated for every fragment contained in it)